app/models/user.rb

```diff
@@ -168,6 +171,29 @@ def change_password(oldpwd, newpwd)
     end
   end
 
+  def locked?
+    ::Settings.authentication.max_failed_login_attempts.positive? && failed_login_attempts >= ::Settings.authentication.max_failed_login_attempts
+  end
+
+  def unlock!
+    update(:failed_login_attempts => 0)
+  end
+
+  def fail_login!
+    update(:failed_login_attempts => failed_login_attempts + 1)
+
+    if locked?
+      queue_opts = {
+        :class_name  => self.class.name,
+        :instance_id => id,
+        :method_name => 'unlock!',
+        :deliver_on  => Time.now.utc + ::Settings.authentication.locked_account_timeout
```

Comment:
I think this call should go on with highest priority.  We don't want the unlock stuck behind a bunch of reports ;)